lib/bolt/catalog.rb

```diff
@@ -54,7 +54,7 @@ def generate_ast(code, filename = nil)
     end
 
     def compile_catalog(request)
-      pal_main = request['code_ast'] || request['code_string']
+      pal_main = request['code_ast'] || request['code_string'] || generate_ast('')
```

Comment:
It seems like this is really inefficient, since Bolt will need to generate the same AST for every target. 

I think it would be better to either catch the `nil` apply body early and generate the empty AST as part of `Bolt::Applicator.apply` or to cache the empty AST as part of `Bolt::Catalog.compile_catalog` so it only gets generated once.

The latter seems easier, but the former maybe seems a little more correct so the catalog input doesn't receive a `nil` value in the first place?

Reply:
I put this here because it ensures that any other tools that call into compile_catalog with a nil string behave the same with Bolt apply blocks. For example, Skeletor may need to write their own applicator that will use the same bolt_catalog task. I'm not sure how real that possibility is though.

Reply:
I guess we can do both, that way this won't call generate_ast for every target but will get called if we ever end up here through other routes: https://github.com/puppetlabs/orchestrator/blob/master/src/puppetlabs/orchestrator/transports/ssh.clj#L87